### Contents

01 Import libraries

02 Import datasets

03 Mixed-type data

04 Missing values

05 Duplicate values

06 Export products dataframe

07 Task 4.5

08 Export orders dataframe

01 Import libraries

In [2]:
import pandas as pd
import numpy as np
import os

02 Import datasets

In [4]:
path = r"C:\Users\cathe\OneDrive\Data Analysis\2 4 Instacart Basket Analysis\02 Data"

In [5]:
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

03 Mixed-Type Data

In [10]:
# Create a dataframe
df_test = pd.DataFrame()

In [12]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [14]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [18]:
df_test.dtypes

mix    object
dtype: object

In [22]:
# Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [24]:
df_test['mix'] = df_test['mix'].astype('str')

04 Missing Values

In [27]:
# Count number of missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [29]:
# View rows containing missing values

df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [31]:
df_prods.shape

(49693, 5)

In [33]:
# Create subset without missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49677, 5)

05 Duplicate values

In [36]:
# Create a subset containing any duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [38]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [40]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [42]:
df_prods_clean_no_dups.shape

(49672, 5)

06 Export products dataframe

In [44]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'))

07 Task

In [9]:
# Consistency checks on orders

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

Comments:
The counts are the same for each column except days_since_prior_order, suggesting there is missing data in this column (about 200,000 rows).
Min order_id is 1 and max is same as count. 50% is half of count.  This all looks expected.
Min user_id is 1 and max is less than order_id.  Fewer users than orders because at least some users have more than one order.  Also to be expected.
Min order_number 1 and max 100.  Interesting.  50% is 11 so most users don't reach 100 orders!
Min order_dow is 0 and max is 6 - as expected.
Min order_hour_of_day is 0 and max is 23 - as expected.
Min days_since_prior_order is 0 and max is 30.  Seems reasonable.

In [51]:
# Check for columns with mixed types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

No mixed type columns

In [11]:
# Count frequency of missing values

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Q5 As noted after running df.describe(), there are 206,209 missing entries in the days_since_prior_order column.  This matches perfectly the maximum user_id, so these rows must be the first orders of each user.

Q6 First I considered imputing the median for the empty cells (not the mean since the mean > median, showing that the data is skewed by extreme values).  However, I struggled to write code to do this without overwriting the dataframe.  On reflection, I decided the empty cells are an indicator of the order being the user's first order (along with order_number == 1), so decided to leave them as they are.

In [68]:
# Check for duplicate values

df_ords_dups = df_ords[df_ords.duplicated()]

In [70]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

Q7  No duplicates found.  If there were, I would remove them, creating a new dataframe without the duplicated rows

08 Export orders dataframe

In [13]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_checked.csv'))